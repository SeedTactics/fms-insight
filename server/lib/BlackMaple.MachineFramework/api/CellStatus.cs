/* Copyright (c) 2023, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#nullable enable

using System;
using System.Runtime.Serialization;
using System.Collections.Immutable;
using Germinate;

namespace BlackMaple.MachineFramework
{
  [DataContract]
  public record QueueSize
  {
    //once an output queue grows to this size, stop unloading parts
    //and keep them in the buffer inside the cell
    [DataMember(IsRequired = false, EmitDefaultValue = false)]
    public int? MaxSizeBeforeStopUnloading { get; init; }
  }

  [DataContract, Draftable]
  public record CurrentStatus
  {
    [DataMember(IsRequired = true)]
    public required DateTime TimeOfCurrentStatusUTC { get; init; }

    [DataMember(Name = "Jobs", IsRequired = true)]
    public required ImmutableDictionary<string, MachineFramework.ActiveJob> Jobs { get; init; }

    [DataMember(Name = "Pallets", IsRequired = true)]
    public required ImmutableDictionary<string, PalletStatus> Pallets { get; init; }

    [DataMember(Name = "Material", IsRequired = true)]
    public required ImmutableList<InProcessMaterial> Material { get; init; }

    [DataMember(Name = "Alarms", IsRequired = true)]
    public required ImmutableList<string> Alarms { get; init; }

    [DataMember(Name = "Queues", IsRequired = true)]
    public required ImmutableDictionary<string, QueueSize> QueueSizes { get; init; }

    // The following is only filled in if the machines move to the load station
    // instead of the pallets moving to the machine
    [DataMember(Name = "MachineLocations", IsRequired = false, EmitDefaultValue = false)]
    public ImmutableList<MachineLocation>? MachineLocations { get; init; }

    public static CurrentStatus operator %(CurrentStatus s, Action<ICurrentStatusDraft> f) => s.Produce(f);
  }

  [DataContract]
  public record JobAndDecrementQuantity
  {
    [DataMember(IsRequired = true)]
    public required long DecrementId { get; init; }

    [DataMember(IsRequired = true)]
    public required string JobUnique { get; init; }

    [DataMember(IsRequired = true)]
    public required DateTime TimeUTC { get; init; }

    [DataMember(IsRequired = true)]
    public required string Part { get; init; }

    [DataMember(IsRequired = true)]
    public required int Quantity { get; init; }
  }

  [DataContract]
  public record HistoricData
  {
    [DataMember(IsRequired = true)]
    public required ImmutableDictionary<string, MachineFramework.HistoricJob> Jobs { get; init; }

    [DataMember(IsRequired = true)]
    public required ImmutableList<SimulatedStationUtilization> StationUse { get; init; }
  }

  [DataContract]
  public record PlannedSchedule
  {
    [DataMember(IsRequired = true)]
    public required string LatestScheduleId { get; init; }

    [DataMember(IsRequired = true)]
    public required ImmutableList<MachineFramework.HistoricJob> Jobs { get; init; }

    [DataMember(IsRequired = true)]
    public required ImmutableDictionary<string, int> ExtraParts { get; init; }

    [DataMember(IsRequired = false)]
    public ImmutableList<Workorder>? CurrentUnfilledWorkorders { get; init; }
  }
}
