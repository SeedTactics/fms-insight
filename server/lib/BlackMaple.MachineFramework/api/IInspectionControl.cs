/* Copyright (c) 2017, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;

namespace BlackMaple.MachineWatchInterface
{
  [Serializable, DataContract]
  public struct InspectCount
  {
    [DataMember(IsRequired = true)] public string Counter;
    [DataMember(IsRequired = true)] public int Value;
    [DataMember(IsRequired = true)] public DateTime LastUTC;
  }

  public interface IInspectionControl : IDisposable
  {
    ///Forces the given materialID to be inspected
    void ForceInspection(long materialID, string inspectionType);

    ///Forces the next piece that runs on the given location to be inspected
    void NextPieceInspection(MachineWatchInterface.PalletLocation palLoc, string inspType);

    //allow editing of counts.  Uses lists of InspectCount structure
    List<InspectCount> LoadInspectCounts();
    void SetInspectCounts(IEnumerable<InspectCount> countUpdates);
  }
}
