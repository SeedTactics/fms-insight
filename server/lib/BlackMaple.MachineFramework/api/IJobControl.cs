/* Copyright (c) 2017, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

using System;
using System.Collections.Generic;

namespace BlackMaple.MachineWatchInterface
{
  public delegate void NewCurrentStatus(CurrentStatus status);

  public interface IJobControl
  {
    ///loads info
    CurrentStatus GetCurrentStatus();

    //checks to see if the jobs are valid.  Some machine types might not support all the different
    //pallet->part->machine->process combinations.
    //Return value is a list of strings, detailing the problems.
    //An empty list or nothing signals the jobs are valid.
    List<string> CheckValidRoutes(IEnumerable<JobPlan> newJobs);

    ///Adds new jobs into the cell controller
    void AddJobs(NewJobs jobs, string expectedPreviousScheduleId);

    void SetJobComment(string jobUnique, string comment);

    //Remove all planned parts from all jobs in the system.
    //
    //The function does 2 things:
    // - Check for planned but not yet machined quantities and if found remove them
    //   and store locally in the machine watch database with a new DecrementId.
    // - Load all decremented quantities (including the potentially new quantities)
    //   strictly after the given decrement ID.
    //Thus this function can be called multiple times to receive the same data.
    List<JobAndDecrementQuantity> DecrementJobQuantites(long loadDecrementsStrictlyAfterDecrementId);
    List<JobAndDecrementQuantity> DecrementJobQuantites(DateTime loadDecrementsAfterTimeUTC);

    //In-process queues

    /// Add new raw material for part.  The part material has not yet been assigned to a specific job,
    /// and will be assigned to the job with remaining demand and earliest priority.
    /// The serial is optional and is passed only if the material has already been marked with a serial.
    void AddUnallocatedPartToQueue(string partName, string queue, int position, string serial);

    /// Add new castings.  The casting has not yet been assigned to a specific job,
    /// and will be assigned to the job with remaining demand and earliest priority.
    /// The serial is optional and is passed only if the material has already been marked with a serial.
    void AddUnallocatedCastingToQueue(string casting, int qty, string queue, int position, IList<string> serial);

    /// Add a new unprocessed piece of material for the given job into the given queue.  The serial is optional
    /// and is passed only if the material has already been marked with a serial.
    /// Use -1 or 0 for lastCompletedProcess if the material is a casting.
    void AddUnprocessedMaterialToQueue(string jobUnique, int lastCompletedProcess, int pathGroup, string queue, int position, string serial);

    /// Add material into a queue or just into free material if the queue name is the empty string.
    /// The material will be inserted into the given position, bumping any later material to a
    /// larger position.  If the material is currently in another queue or a different position,
    /// it will be removed and placed in the given position.
    void SetMaterialInQueue(long materialId, string queue, int position);

    void RemoveMaterialFromAllQueues(IList<long> materialIds);

    event NewCurrentStatus OnNewCurrentStatus;
  }

  public interface IOldJobDecrement
  {
    //The old method of decrementing, which stores only a single decrement until finalize is called.
    Dictionary<JobAndPath, int> OldDecrementJobQuantites();
    void OldFinalizeDecrement();
  }
}