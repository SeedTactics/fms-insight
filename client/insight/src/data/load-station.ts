/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import { HashMap, Vector } from "prelude-ts";

export interface PalletData {
  pallet: api.IPalletStatus;
  material: api.IInProcessMaterial[];
}

export function buildPallets(
  st: Readonly<api.ICurrentStatus>
): HashMap<string, { pal?: PalletData; queued?: PalletData }> {
  const matByPallet = new Map<string, api.IInProcessMaterial[]>();
  for (const mat of st.material) {
    if (mat.location.type === api.LocType.OnPallet && mat.location.pallet !== undefined) {
      const mats = matByPallet.get(mat.location.pallet) || [];
      mats.push(mat);
      matByPallet.set(mat.location.pallet, mats);
    }
  }

  const m = new Map<string, { pal?: PalletData; queued?: PalletData }>();
  for (const pal of Object.values(st.pallets)) {
    switch (pal.currentPalletLocation.loc) {
      case api.PalletLocationEnum.LoadUnload:
      case api.PalletLocationEnum.Machine: {
        const stat = pal.currentPalletLocation.group + " #" + pal.currentPalletLocation.num.toString();
        m.set(stat, {
          ...(m.get(stat) || {}),
          pal: {
            pallet: pal,
            material: matByPallet.get(pal.pallet) || [],
          },
        });
        break;
      }

      case api.PalletLocationEnum.MachineQueue: {
        const stat2 = pal.currentPalletLocation.group + " #" + pal.currentPalletLocation.num.toString();
        m.set(stat2, {
          ...(m.get(stat2) || {}),
          queued: {
            pallet: pal,
            material: matByPallet.get(pal.pallet) || [],
          },
        });
        break;
      }

      // TODO: buffer and cart
    }
  }

  return HashMap.ofIterable(m);
}

export type MaterialList = ReadonlyArray<Readonly<api.IInProcessMaterial>>;

export interface LoadStationAndQueueData {
  readonly loadNum: number;
  readonly pallet?: Readonly<api.IPalletStatus>;
  readonly face: HashMap<number, MaterialList>;
  readonly stationStatus?: HashMap<string, { pal?: PalletData; queued?: PalletData }>;
  readonly freeLoadingMaterial: MaterialList;
  readonly free?: MaterialList;
  readonly queues: HashMap<string, MaterialList>;
}

export function selectLoadStationAndQueueProps(
  loadNum: number,
  queues: ReadonlyArray<string>,
  displayFree: boolean,
  curSt: Readonly<api.ICurrentStatus>
): LoadStationAndQueueData {
  let pal: Readonly<api.IPalletStatus> | undefined;
  if (loadNum >= 0) {
    for (const p of Object.values(curSt.pallets)) {
      if (
        p.currentPalletLocation.loc === api.PalletLocationEnum.LoadUnload &&
        p.currentPalletLocation.num === loadNum
      ) {
        pal = p;
        break;
      }
    }
  }

  let byFace: HashMap<number, api.IInProcessMaterial[]> = HashMap.empty();
  let palName: string | undefined;
  let freeLoading: ReadonlyArray<Readonly<api.IInProcessMaterial>> = [];
  let stationStatus: HashMap<string, { pal?: PalletData; queued?: PalletData }> | undefined;

  // first free and queued material
  let free: MaterialList | undefined;
  if (displayFree) {
    free = curSt.material.filter(
      (m) => m.action.processAfterLoad && m.action.processAfterLoad > 1 && m.location.type === api.LocType.Free
    );
  }

  const queueNames = HashMap.ofIterable<string, api.IInProcessMaterial[]>(
    queues.map((q) => [q, []] as [string, api.IInProcessMaterial[]])
  );
  const queueMat = Vector.ofIterable(curSt.material)
    .filter((m) => {
      if (m.location.type !== api.LocType.InQueue) {
        return false;
      }
      if (queueNames.containsKey(m.location.currentQueue || "")) {
        return true;
      } else if (pal && m.action.type === api.ActionType.Loading && m.action.loadOntoPallet === pal.pallet) {
        return true; // display even if queue not selected
      }
      return false;
    })
    .groupBy((m) => m.location.currentQueue || "")
    .mapValues((ms) => ms.sortOn((m) => m.location.queuePosition || 0).toArray());

  // load pallet material
  if (pal !== undefined) {
    palName = pal.pallet;

    byFace = Vector.ofIterable(curSt.material)
      .filter((m) => m.location.type === api.LocType.OnPallet && m.location.pallet === palName)
      .groupBy((m) => m.location.face || 0)
      .mapValues((ms) => ms.toArray());

    freeLoading = curSt.material.filter(
      (m) =>
        m.action.type === api.ActionType.Loading &&
        m.action.loadOntoPallet === palName &&
        m.action.processAfterLoad === 1 &&
        m.location.type === api.LocType.Free
    );

    // make sure all face desinations exist
    queueMat.forEach(([_, mats]) =>
      mats.forEach((m) => {
        if (m.action.type === api.ActionType.Loading && m.action.loadOntoPallet === palName) {
          const face = m.action.loadOntoFace;
          if (face !== undefined && !byFace.containsKey(face)) {
            byFace = byFace.put(face, []);
          }
        }
      })
    );
    [...freeLoading, ...(free || [])].forEach((m) => {
      if (m.action.type === api.ActionType.Loading && m.action.loadOntoPallet === palName) {
        const face = m.action.loadOntoFace;
        if (face !== undefined && !byFace.containsKey(face)) {
          byFace = byFace.put(face, []);
        }
      }
    });
  } else {
    stationStatus = buildPallets(curSt);
    if (displayFree) {
      freeLoading = curSt.material.filter(
        (m) => m.action.processAfterLoad === 1 && m.location.type === api.LocType.Free
      );
    }
  }

  return {
    loadNum,
    pallet: pal,
    face: byFace,
    stationStatus: stationStatus,
    freeLoadingMaterial: freeLoading,
    free: free,
    queues: queueNames.mergeWith(queueMat, (m1, m2) => [...m1, ...m2]),
  };
}
