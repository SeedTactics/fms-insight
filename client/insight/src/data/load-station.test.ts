/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import curSt from "../../test/status-mock.json";
import { CurrentStatus } from "../network/api.js";
import { selectLoadStationAndQueueProps } from "./load-station.js";
import { describe, it, expect } from "vitest";
import { toRawJs } from "../../test/to-raw-js.js";

describe("load station status", () => {
  it("load 1 with no queues", () => {
    const status = CurrentStatus.fromJS(curSt);
    expect(toRawJs(selectLoadStationAndQueueProps(1, [], status))).toMatchSnapshot("load 1 with no queues");
  });

  it("load 2 with queue", () => {
    const status = CurrentStatus.fromJS(curSt);
    expect(toRawJs(selectLoadStationAndQueueProps(2, ["Queue1"], status))).toMatchSnapshot(
      "load 2 with queue"
    );
  });

  it("load 3 with empty pallet", () => {
    const status = CurrentStatus.fromJS(curSt);
    expect(toRawJs(selectLoadStationAndQueueProps(3, ["Queue2"], status))).toMatchSnapshot(
      "load 3 with empty pallet"
    );
  });
});
