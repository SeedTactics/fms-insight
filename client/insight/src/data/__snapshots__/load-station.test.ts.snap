// Vitest Snapshot v1

exports[`load station status > load 1 with no queues > load 1 with no queues 1`] = `
{
  "face": Map {
    1 => [],
  },
  "freeLoadingMaterial": [
    {
      "action": {
        "elapsedLoadUnloadTime": undefined,
        "elapsedMachiningTime": undefined,
        "expectedRemainingMachiningTime": undefined,
        "loadOntoFace": 1,
        "loadOntoPallet": "1",
        "pathAfterLoad": 1,
        "processAfterLoad": 1,
        "program": undefined,
        "type": "Loading",
        "unloadIntoQueue": undefined,
      },
      "jobUnique": "aaa-offline-2018-01-01",
      "lastCompletedMachiningRouteStopIndex": undefined,
      "location": {
        "currentQueue": undefined,
        "face": undefined,
        "pallet": undefined,
        "queuePosition": undefined,
        "type": "Free",
      },
      "materialID": 100663,
      "partName": "bbb",
      "path": 1,
      "process": 1,
      "serial": "NRWGM",
      "signaledInspections": [],
      "workorderId": undefined,
    },
  ],
  "loadNum": 1,
  "pallet": {
    "currentPalletLocation": {
      "group": "L/U",
      "loc": "LoadUnload",
      "num": 1,
    },
    "fixtureOnPallet": "fix1",
    "newFixture": undefined,
    "numFaces": 2,
    "onHold": false,
    "pallet": "1",
    "percentMoveCompleted": undefined,
    "targetLocation": undefined,
  },
  "queues": Map {},
  "stationStatus": undefined,
}
`;

exports[`load station status > load 2 with queue > load 2 with queue 1`] = `
{
  "face": Map {
    1 => [],
  },
  "freeLoadingMaterial": [],
  "loadNum": 2,
  "pallet": {
    "currentPalletLocation": {
      "group": "L/U",
      "loc": "LoadUnload",
      "num": 2,
    },
    "fixtureOnPallet": "fix5",
    "newFixture": undefined,
    "numFaces": 1,
    "onHold": false,
    "pallet": "5",
    "percentMoveCompleted": undefined,
    "targetLocation": undefined,
  },
  "queues": Map {
    "Queue1" => [
      {
        "action": {
          "elapsedLoadUnloadTime": undefined,
          "elapsedMachiningTime": undefined,
          "expectedRemainingMachiningTime": undefined,
          "loadOntoFace": undefined,
          "loadOntoPallet": undefined,
          "pathAfterLoad": undefined,
          "processAfterLoad": undefined,
          "program": undefined,
          "type": "Waiting",
          "unloadIntoQueue": undefined,
        },
        "jobUnique": "aaa-offline-2018-01-01",
        "lastCompletedMachiningRouteStopIndex": undefined,
        "location": {
          "currentQueue": "Queue1",
          "face": undefined,
          "pallet": undefined,
          "queuePosition": 0,
          "type": "InQueue",
        },
        "materialID": 104289,
        "partName": "aaa",
        "path": 1,
        "process": 1,
        "serial": "MRYHE",
        "signaledInspections": [],
        "workorderId": undefined,
      },
      {
        "action": {
          "elapsedLoadUnloadTime": undefined,
          "elapsedMachiningTime": undefined,
          "expectedRemainingMachiningTime": undefined,
          "loadOntoFace": undefined,
          "loadOntoPallet": undefined,
          "pathAfterLoad": undefined,
          "processAfterLoad": undefined,
          "program": undefined,
          "type": "Waiting",
          "unloadIntoQueue": undefined,
        },
        "jobUnique": "xxx-offline-2018-01-01",
        "lastCompletedMachiningRouteStopIndex": undefined,
        "location": {
          "currentQueue": "Queue1",
          "face": undefined,
          "pallet": undefined,
          "queuePosition": 1,
          "type": "InQueue",
        },
        "materialID": 104458,
        "partName": "xxx",
        "path": 1,
        "process": 1,
        "serial": "NBVHA",
        "signaledInspections": [],
        "workorderId": undefined,
      },
      {
        "action": {
          "elapsedLoadUnloadTime": undefined,
          "elapsedMachiningTime": undefined,
          "expectedRemainingMachiningTime": undefined,
          "loadOntoFace": 1,
          "loadOntoPallet": "5",
          "pathAfterLoad": 1,
          "processAfterLoad": 1,
          "program": undefined,
          "type": "Loading",
          "unloadIntoQueue": undefined,
        },
        "jobUnique": "zzz-offline-2018-01-01",
        "lastCompletedMachiningRouteStopIndex": undefined,
        "location": {
          "currentQueue": "Queue1",
          "face": undefined,
          "pallet": undefined,
          "queuePosition": 2,
          "type": "InQueue",
        },
        "materialID": 102304,
        "partName": "zzz",
        "path": 1,
        "process": 1,
        "serial": "MZQGQ",
        "signaledInspections": [],
        "workorderId": undefined,
      },
    ],
  },
  "stationStatus": undefined,
}
`;

exports[`load station status > load 3 with empty pallet > load 3 with empty pallet 1`] = `
{
  "face": Map {},
  "freeLoadingMaterial": [],
  "loadNum": 3,
  "pallet": undefined,
  "queues": Map {
    "Queue2" => [
      {
        "action": {
          "elapsedLoadUnloadTime": undefined,
          "elapsedMachiningTime": undefined,
          "expectedRemainingMachiningTime": undefined,
          "loadOntoFace": undefined,
          "loadOntoPallet": undefined,
          "pathAfterLoad": undefined,
          "processAfterLoad": undefined,
          "program": undefined,
          "type": "Waiting",
          "unloadIntoQueue": undefined,
        },
        "jobUnique": "ccc-offline-2018-01-01",
        "lastCompletedMachiningRouteStopIndex": undefined,
        "location": {
          "currentQueue": "Queue2",
          "face": undefined,
          "pallet": undefined,
          "queuePosition": 0,
          "type": "InQueue",
        },
        "materialID": 104824,
        "partName": "ccc",
        "path": 1,
        "process": 1,
        "serial": "OBSGQ",
        "signaledInspections": [],
        "workorderId": undefined,
      },
      {
        "action": {
          "elapsedLoadUnloadTime": undefined,
          "elapsedMachiningTime": undefined,
          "expectedRemainingMachiningTime": undefined,
          "loadOntoFace": undefined,
          "loadOntoPallet": undefined,
          "pathAfterLoad": undefined,
          "processAfterLoad": undefined,
          "program": undefined,
          "type": "Waiting",
          "unloadIntoQueue": undefined,
        },
        "jobUnique": "zzz-offline-2018-01-01",
        "lastCompletedMachiningRouteStopIndex": undefined,
        "location": {
          "currentQueue": "Queue2",
          "face": undefined,
          "pallet": undefined,
          "queuePosition": 1,
          "type": "InQueue",
        },
        "materialID": 105125,
        "partName": "zzz",
        "path": 1,
        "process": 1,
        "serial": "MJSGU",
        "signaledInspections": [],
        "workorderId": undefined,
      },
    ],
  },
  "stationStatus": Map {
    "L/U #1" => {
      "pal": {
        "material": [],
        "pallet": {
          "currentPalletLocation": {
            "group": "L/U",
            "loc": "LoadUnload",
            "num": 1,
          },
          "fixtureOnPallet": "fix1",
          "newFixture": undefined,
          "numFaces": 2,
          "onHold": false,
          "pallet": "1",
          "percentMoveCompleted": undefined,
          "targetLocation": undefined,
        },
      },
    },
    "Machine #1" => {
      "pal": {
        "material": [],
        "pallet": {
          "currentPalletLocation": {
            "group": "Machine",
            "loc": "Machine",
            "num": 1,
          },
          "fixtureOnPallet": "fix4",
          "newFixture": undefined,
          "numFaces": 2,
          "onHold": false,
          "pallet": "4",
          "percentMoveCompleted": undefined,
          "targetLocation": undefined,
        },
      },
    },
    "L/U #2" => {
      "pal": {
        "material": [],
        "pallet": {
          "currentPalletLocation": {
            "group": "L/U",
            "loc": "LoadUnload",
            "num": 2,
          },
          "fixtureOnPallet": "fix5",
          "newFixture": undefined,
          "numFaces": 1,
          "onHold": false,
          "pallet": "5",
          "percentMoveCompleted": undefined,
          "targetLocation": undefined,
        },
      },
    },
    "Machine #2" => {
      "pal": {
        "material": [],
        "pallet": {
          "currentPalletLocation": {
            "group": "Machine",
            "loc": "Machine",
            "num": 2,
          },
          "fixtureOnPallet": "fix6",
          "newFixture": undefined,
          "numFaces": 1,
          "onHold": false,
          "pallet": "6",
          "percentMoveCompleted": undefined,
          "targetLocation": undefined,
        },
      },
      "queued": {
        "material": [],
        "pallet": {
          "currentPalletLocation": {
            "group": "Machine",
            "loc": "MachineQueue",
            "num": 2,
          },
          "fixtureOnPallet": "fix7",
          "newFixture": undefined,
          "numFaces": 1,
          "onHold": false,
          "pallet": "7",
          "percentMoveCompleted": undefined,
          "targetLocation": undefined,
        },
      },
    },
    "Machine #3" => {
      "pal": {
        "material": [],
        "pallet": {
          "currentPalletLocation": {
            "group": "Machine",
            "loc": "Machine",
            "num": 3,
          },
          "fixtureOnPallet": "fix8",
          "newFixture": undefined,
          "numFaces": 1,
          "onHold": false,
          "pallet": "8",
          "percentMoveCompleted": undefined,
          "targetLocation": undefined,
        },
      },
    },
  },
}
`;
